internal/worker/objectstores3caller/manifold.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/juju/juju/internal/s3client"
+	"github.com/juju/juju/internal/servicefactory"
+)
+
+// Logger represents the logging methods called.
+type Logger interface {
+	Errorf(message string, args ...any)
+	Warningf(message string, args ...any)
+	Infof(message string, args ...any)
+	Debugf(message string, args ...any)
+	Tracef(message string, args ...any)
+
+	IsTraceEnabled() bool
+}
+
+// NewClientFunc is a function that returns a new S3 client.
+type NewClientFunc func(string, s3client.HTTPClient, s3client.Credentials, s3client.Logger) (objectstore.Session, error)
+
+// GetControllerConfigServiceFunc is a helper function that gets a service from
+// the manifold.
+type GetControllerConfigServiceFunc func(getter dependency.Getter, name string) (ControllerService, error)
+
+// NewWorkerFunc is a function that returns a new worker.
+type NewWorkerFunc func(workerConfig) (worker.Worker, error)
+
+// ManifoldConfig defines a Manifold's dependencies.
+type ManifoldConfig struct {
+	HTTPClientName     string
+	ServiceFactoryName string
+
+	// NewClient is used to create a new object store client.
+	NewClient NewClientFunc
+	// Logger is used to write logging statements for the worker.
+	Logger Logger
+	// Clock is used for the retry mechanism.
+	Clock clock.Clock
+
+	// GetControllerConfigService is used to get a service from the manifold.
+	GetControllerConfigService GetControllerConfigServiceFunc
+	NewWorker                  NewWorkerFunc
+}
+
+func (cfg ManifoldConfig) Validate() error {
+	if cfg.HTTPClientName == "" {
+		return errors.NotValidf("nil HTTPClientName")
+	}
+	if cfg.ServiceFactoryName == "" {
+		return errors.NotValidf("nil ServiceFactoryName")
+	}
+	if cfg.NewClient == nil {
+		return errors.NotValidf("nil NewClient")
+	}
+	if cfg.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if cfg.Clock == nil {
+		return errors.NotValidf("nil Clock")
+	}
+	return nil
+}
+
+// Manifold returns a manifold whose worker wraps an S3 Session.
+func Manifold(config ManifoldConfig) dependency.Manifold {
+	return dependency.Manifold{
+		Inputs: []string{
+			config.HTTPClientName,
+			config.ServiceFactoryName,
+		},
+		Output: outputFunc,
+		Start:  config.start,
+	}
+}
+
+// start returns a StartFunc that creates a S3 client based on the supplied
+// manifold config and wraps it in a worker.
+func (config ManifoldConfig) start(ctx context.Context, getter dependency.Getter) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	controllerConfigService, err := config.GetControllerConfigService(getter, config.ServiceFactoryName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	controllerConfig, err := controllerConfigService.ControllerConfig(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// If we're not using S3, then we don't need to start this worker.
+	if controllerConfig.ObjectStoreType() != objectstore.S3Backend {
```

Comment:
We can do this in the manifold because we don't allow updates to this article.

But that means you _must_ know in advance that you will go into HA at some stage in order to use S3 from the outset right? Because file-based is HA-1 only.

Reply:
> But that means you must know in advance that you will go into HA at some stage in order to use S3 from the outset right?

Yes. This is a current restriction. This will be removed in the future when we allow draining from file to s3, Until that time, there is no point having a worker running if you can't use it.